File: backend/server/src/main/java/com/groudina/ten/demo/jwt/JWTProvider.java
package com.groudina.ten.demo.jwt;

import com.groudina.ten.demo.datasource.DbUserRepository;
import io.jsonwebtoken.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.Serializable;
import java.util.Date;

@Component
public class JWTProvider {
    private final Logger logger = LoggerFactory.getLogger(JWTProvider.class);

    @Value("${groudina.app.jwtSecret}")
    private String jwtSecret;

    @Value("${groudina.app.jwtExpiration}")
    private Integer jwtExpiration = 0;

    public String generateJwtToken(String email) {
        return Jwts.builder()
                .setSubject(email)
                .setIssuedAt(new Date())
                .setExpiration(new Date(new Date().getTime() + jwtExpiration * 1000))
                .signWith(SignatureAlgorithm.HS512, jwtSecret)
                .compact();
    }

    public boolean validateJwtToken(String authToken) {
        try {
            Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(authToken);
            return true;
        } catch (SignatureException e) {
            logger.error(String.format("Invalid JwtSignature: error: %s", e.getMessage()));
        } catch (MalformedJwtException e) {
            logger.error(String.format("Invalid JwtToken: error: %s", e.getMessage()));
        } catch (ExpiredJwtException e) {
            logger.error(String.format("Expired JWT token: error: %s", e.getMessage()));
        } catch (UnsupportedJwtException e) {
            logger.error(String.format("Unsupported JWT token: error: %s", e.getMessage()));
        } catch (IllegalArgumentException e) {
            logger.error(String.format("JWT claims string is empty: error: %s", e.getMessage()));
        }

        return false;
    }

    public String getEmailFromJwtToken(String token) {
        return Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(token).getBody().getSubject();
    }
}

File: backend/app/web/src/main/kotlin/ru/nekoguys/game/web/security/jwt/JwtProvider.kt
package ru.nekoguys.game.web.security.jwt

import io.jsonwebtoken.*
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.SecurityException
import org.slf4j.LoggerFactory
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtProvider(
    private val properties: JwtProperties,
) {
    private val logger = LoggerFactory.getLogger(JwtProvider::class.java)

    private val jwtSecretKey: SecretKey =
        calculateSecretKey(properties.secret)

    private fun calculateSecretKey(secret: String?): SecretKey =
        if (secret.isNullOrBlank()) {
            Keys.secretKeyFor(SignatureAlgorithm.HS256)
        } else {
            Keys.hmacShaKeyFor(secret.toByteArray())
        }

    private val jwtParser: JwtParser =
        Jwts.parserBuilder().setSigningKey(jwtSecretKey).build()

    fun generateExpirationTimestamp(): Long =
        Date().time + properties.expirationSeconds * 1000

    fun generateJwtToken(authentication: Authentication, expirationTimestamp: Long): String {
        val roles = authentication
            .authorities
            .joinToString(";") { "$it".removePrefix("ROLE_") }

        return Jwts.builder()
            .setSubject(authentication.name)
            .claim("roles", roles)
            .setIssuedAt(Date()) // не совсем корректный способ
            .setExpiration(Date(expirationTimestamp))
            .signWith(jwtSecretKey)
            .compact()
    }

    fun parseJwt(token: String): Jws<Claims>? {
        try {
            return jwtParser.parseClaimsJws(token)
        } catch (e: SecurityException) {
            logger.debug("Invalid JwtSignature", e)
        } catch (e: MalformedJwtException) {
            logger.debug("Invalid JwtToken", e)
        } catch (e: ExpiredJwtException) {
            logger.debug("Expired JWT token", e)
        } catch (e: UnsupportedJwtException) {
            logger.debug("Unsupported JWT token", e)
        } catch (e: IllegalArgumentException) {
            logger.debug("JWT claims string $token is empty", e)
        }
        return null
    }
}
